RAII chapter for idiomatic rust (#2820)
google/comprehensive-rust: src/idiomatic/leveraging-the-type-system/raii.md

  trait to release the resource automatically. This ties cleanup to the lifetime
  of the `File` value.

  ```rust,compile_fail
  impl Drop for File {
      fn drop(&mut self) {
          println!("release file descriptor automatically");
      }
  }
  ```

- Note that `Drop::drop` cannot return errors. Any fallible logic must be
  handled internally or ignored. In the standard library, errors returned while
  closing an owned file descriptor during `Drop` are silently discarded:
  <https://doc.rust-lang.org/src/std/os/fd/owned.rs.html#169-196>

- If both `drop()` and `close()` exist, the file descriptor may be released
  twice. To avoid this, remove `close()` and rely solely on `Drop`.

- When is `Drop::drop` called?

  Normally, when the `file` variable in `main` goes out of scope (either on
  return or due to a panic), `drop()` is called automatically.

  If the file is moved into another function, for example `read_all()`, the
  value is dropped when that function returns — not in `main`.

  In contrast, C++ runs destructors in the original scope even for moved-from
  values.

- The same mechanism powers `std::mem::drop`:

  ```rust
  pub fn drop<T>(_x: T) {}

>   ```

Reviewer: This got lost.

>   ```

Author: This is no longer anywhere in the slides. Given all the content we have now it didn't seem to add any value.

  of the `File` value.

  ```rust,compile_fail
  impl Drop for File {
      fn drop(&mut self) {
          println!("release file descriptor automatically");
      }
  }
  ```

- Note that `Drop::drop` cannot return errors. Any fallible logic must be
  handled internally or ignored. In the standard library, errors returned while
  closing an owned file descriptor during `Drop` are silently discarded:
  <https://doc.rust-lang.org/src/std/os/fd/owned.rs.html#169-196>

- If both `drop()` and `close()` exist, the file descriptor may be released
  twice. To avoid this, remove `close()` and rely solely on `Drop`.

- When is `Drop::drop` called?

  Normally, when the `file` variable in `main` goes out of scope (either on
  return or due to a panic), `drop()` is called automatically.

  If the file is moved into another function, for example `read_all()`, the
  value is dropped when that function returns — not in `main`.

  In contrast, C++ runs destructors in the original scope even for moved-from
  values.

- The same mechanism powers `std::mem::drop`:

  ```rust
  pub fn drop<T>(_x: T) {}
  ```

  You can use it to force early destruction of a value before its natural end of
  scope.

> - Insert `panic!("oops")` at the start of `read_to_end()` to show that `drop()`
>   still runs during unwinding. Rust guarantees this unless the panic strategy is
>   set to `abort`.

Reviewer 2: ```suggestion
- Insert `panic!("oops")` at the start of `read_to_end()` to show that `drop()`
  still runs during unwinding.
```

You cover this in the next bullet point, so I think this one can be simplified.

---
minutes: 30
---

# RAII and `Drop` in Practice

RAII (_Resource Acquisition Is Initialization_) means tying the lifetime of a
resource to the lifetime of a value.

> Rust applies RAII automatically for memory management. The `Drop` trait lets you

Author: Thank you. I admire your skill to find the most precise language a lot. It's clear I'll learn a lot from you on this project.

extend this pattern to anything else.

```rust
use std::sync::Mutex;

fn main() {
    let mux = Mutex::new(vec![1, 2, 3]);

    {
        let mut data = mux.lock().unwrap();
        data.push(4); // lock held here
    } // lock automatically released here

> }

Author: I like this flow a lot. Seems I can do more slides than I thought. Would you still keep the slide about scopeguard at the end?

> }

Reviewer: Yes, absolutely, it should be a part of the explanation of how Mutex uses RAII.

```

<details>

- In the above example
  [the `Mutex`](https://doc.rust-lang.org/std/sync/struct.Mutex.html) owns its
  data: you can’t access the value inside without first acquiring the lock.

  `mux.lock()` returns a
  [`MutexGuard`](https://doc.rust-lang.org/std/sync/struct.MutexGuard.html),
  which [dereferences](https://doc.rust-lang.org/std/ops/trait.DerefMut.html) to
  the data and implements

>   [`Drop`](https://doc.rust-lang.org/std/ops/trait.Drop.html).

Author: Given this and your other comment about not being sure about whether a topic like Mutex even belongs here, do you still think we should devote these 2 extra slides to it? And if so, where do you see this in the flow?

- You may recall from

>   [the Memory Management chapter](../../memory-management/drop.md) that the

Author: For links you might want to use a tool like Lychee, https://github.com/lycheeverse/lychee. It makes it trivial to auto-detect broken links.

There are also seem to be mdbook specific tooling, but I do not have experience with those: https://github.com/Michael-F-Bryan/mdbook-linkcheck

  [`Drop` trait](https://doc.rust-lang.org/std/ops/trait.Drop.html) lets you
  define what should happen when a resource is dropped.

>   - In

Reviewer 3: It's probably only necessary to include one "callback" to Fundamentals -- the important point is to that this slide is a quick review of previous content, and if students need a deeper refresher they can find that content in the Fundamentals course.

That said, these speaker notes are pretty long! Is it possible to trim this down to just call out the bits necessary for the RAII patterns introduced here, leaving the rest to the students' memory of Fundamentals?

    [the Blocks and Scopes chapter](../../control-flow-basics/blocks-and-scopes.md),
    we saw the most common situation where a resource is dropped: when the scope
    of its _owner_ ends at the boundary of a block (`{}`).

  - The use of
    [`std::mem::drop(val)`](https://doc.rust-lang.org/std/mem/fn.drop.html)
    allows you to _move_ a value out of scope before the block ends.

  - There are also other scenarios where this can happen, such as when the value
    owning the resource is "shadowed" by another value:

    ```rust
    let a = String::from("foo");
    let a = 3; // ^ The previous string is dropped here

>                //   because we shadow its binding with a new value.

Reviewer 2: Shadowing DOES NOT impact drop, a shadowed variable remains in place and still goes out of scope at the same time.

I'd suggest adding this as a "More to Know" point at the bottom of the speaker notes. This is a nice detail to know, but it probably only needs to be covered if students ask about it.

    ```

  - Recall also from [the Drop chapter](../../memory-management/drop.md) that
    for a composite type such as a `struct`, all its fields will be dropped when
    the struct itself is dropped. If a field implements the `Drop` trait, its
    `Drop::drop` _trait_ method will also be invoked.

> - In any scenario where the stack unwinds the value, it is guaranteed that the

Reviewer 3: The stack doesn't unwind things -- the stack _is_ unwound in handling a panic or normal function return.

    - Returning early with an explicit `return` statement, or implicitly by
      using [the Try operator (`?`)](../../error-handling/try.md) to
      early-return `Option` or `Result` values.

  - It also holds for unexpected scenarios where a `panic` is triggered, if:

    - The stack unwinds on panic (which is the default), allowing for graceful
      cleanup of resources.

      This unwind behavior can be overridden to instead
      [abort on panic](https://github.com/rust-lang/rust/blob/master/library/panic_abort/src/lib.rs).

    - No panic occurs within any of the `drop` methods invoked before reaching
      the `drop` call of the object `a`.

  - Note that
    [an explicit exit of the program](https://doc.rust-lang.org/std/process/fn.exit.html),
    as sometimes used in CLI tools, terminates the process immediately. In other
    words, the stack is not unwound in this case, and the `drop` method will not
    be called.

- `Drop` is a great fit for use cases like `Mutex`.

  When the guard goes out of scope,
  [`Drop::drop`](https://doc.rust-lang.org/std/ops/trait.Drop.html#tymethod.drop)
  is called and unlocks the mutex automatically.

>   In contrast to C++ or Java, where you often have to unlock manually or use a

Author: Yes, that's also where many of us learned about these concepts. Did not mean to imply that. It was mostly about the lock. Your points are well taken though, gonna revisit this with care.

>   In contrast to C++ or Java, where you often have to unlock manually or use a

Reviewer: Yeah, sorry, I was a bit terse in my comment. Even for locks though, high-quality APIs for locks in C++ and Java today are RAII-based ([std::lock_guard](https://en.cppreference.com/w/cpp/thread/lock_guard.html), [absl::MutexLock](https://github.com/abseil/abseil-cpp/blob/master/absl/synchronization/mutex.h#L583), `synchronized(obj) {}` in Java).

  `lock/unlock` pattern, Rust ensures the lock _cannot_ be forgotten, thanks to

>   the compiler.

Author: Great suggestion. Will incorporate this feedback.

- Instead of relying on the user to remember to call `close()`, we can implement
  the `Drop` trait to release the resource automatically. This ties cleanup to
  the lifetime of the `File` value. Note that `Drop` cannot return errors, so
  any fallible logic must be handled internally or avoided.

  ```rust,compile_fail
  impl Drop for FileLock {
      fn drop(&mut self) {
          println!("release file descriptor automatically");
      }
  }
  ```

- If both `drop()` and `close()` are present, the file descriptor is released
  twice. To avoid this, remove `close()` and rely solely on `Drop`.

  This also illustrates that when a parent type is dropped, the `drop()` method
  of its fields (such as `FileLock`) is automatically called — no extra code is
  needed.

- Demonstrate ownership transfer by moving the file into a `read_all()`
  function. The file is dropped when the local variable inside that function
  goes out of scope, not in `main`.

  This differs from C++, where destructors are tied to the original scope, even
  for moved-from values.

  The same mechanism underlies `std::mem::drop`, which lets you drop a value
  early:

  ```rust
  pub fn drop<T>(_x: T) {}

>   ```

Reviewer 4: I get _what_ we want to say, but we should lead with it: _when_ does `Drop::drop` get invoked?
And then elaborate from there.

- Insert `panic!("oops")` at the start of `read_all()` to show that `drop()` is
  still called during unwinding. Rust ensures this unless the panic strategy is
  set to `abort`.

- There are exceptions where destructors will not run:
  - If a destructor panics during unwinding, the program aborts immediately.
  - The program also aborts when using `std::process::exit()` or when compiled
    with the `abort` panic strategy.

### More to Explore

The `Drop` trait has another important limitation: it is not `async`.

You cannot `await` inside a destructor, which is often needed when cleaning up
asynchronous resources like sockets, database connections, or tasks that must
signal completion to another system.

- Learn more:
  <https://rust-lang.github.io/async-fundamentals-initiative/roadmap/async_drop.html>

> - Available on nightly:

Reviewer 4: ```suggestion
- There is an experimental `AsyncDrop` trait in `nightly`:
```

I'd use a more cautious language here, since I don't think there's any path to stabilisation for `AsyncDrop` at the moment.

and the `Drop` trait allows you to extend this to other resources, such as file
descriptors or locks.

```rust,editable
pub struct File(std::os::fd::RawFd);

impl File {
    pub fn open(path: &str) -> Result<Self, std::io::Error> {
        // [...]
        Ok(Self(0))
    }

    pub fn read_to_end(&mut self) -> Result<Vec<u8>, std::io::Error> {
        // [...]
        Ok(b"example".to_vec())
    }

    pub fn close(self) -> Result<(), std::io::Error> {
        // [...]
        Ok(())
    }
}

fn main() -> Result<(), std::io::Error> {
    let mut file = File::open("example.txt")?;
    println!("content: {:?}", file.read_to_end()?);
    Ok(())
}
```

<details>

- Easy to miss: `file.close()` is never called. Ask the class if they noticed.

- To release the file descriptor correctly, `file.close()` must be called after
  the last use — and also in early-return paths in case of errors.

- Instead of relying on the user to call `close()`, we can implement the `Drop`
  trait to release the resource automatically. This ties cleanup to the lifetime
  of the `File` value.

  ```rust,compile_fail
  impl Drop for File {
      fn drop(&mut self) {

>           println!("release file descriptor automatically");

Reviewer: ```suggestion
          // libc::close(...);
          println!("file descriptor was closed");
```

      }
  }
  ```

> - Note that `Drop::drop` cannot return a Result. Any fallible logic must be

Reviewer: ```suggestion
- Note that `Drop::drop()` cannot return a `Result`. Any failures must be
```
